interceptors_node_publicist.go

```diff
@@ -0,0 +1,397 @@
+		}
+		return &csi.NodeUnpublishVolumeResponse{}, nil
+	}
+	return handler(ctx, req)
+}
+
+func (i *nodeVolumePublicist) publish(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest) error {
+
+	if req.VolumeId == "" {
+		return ErrVolumeIDRequired
+	}
+
+	if req.TargetPath == "" {
+		return ErrTargetPathRequired
+	}
+
+	if req.VolumeCapability == nil {
+		return ErrVolumeCapabilityRequired
+	}
+
+	if req.VolumeCapability.AccessMode == nil {
+		return ErrAccessModeRequired
+	}
+
+	var (
+		isBlock    bool
+		fsType     string
+		mntFlags   []string
+		readOnly   = req.Readonly
+		accessMode = req.VolumeCapability.AccessMode.Mode
+	)
+
+	lf := log.Fields{
+		"volumeID":   req.VolumeId,
+		"targetPath": req.TargetPath,
+		"readOnly":   req.Readonly,
+		"accessMode": accessMode,
+	}
+
+	switch accessType := req.VolumeCapability.AccessType.(type) {
+	case *csi.VolumeCapability_Block:
+		if accessType.Block == nil {
+			return ErrBlockTypeRequired
+		}
+		if readOnly {
+			return status.Error(
+				codes.InvalidArgument,
+				"read only not supported by access type")
+		}
+		isBlock = true
+		lf["accessType"] = "block"
+	case *csi.VolumeCapability_Mount:
+		if accessType.Mount == nil {
+			return ErrMountTypeRequired
+		}
+		fsType = accessType.Mount.FsType
+		mntFlags = accessType.Mount.MountFlags
+		lf["accessType"] = "mount"
+		lf["fsType"] = fsType
+		lf["mntFlags"] = mntFlags
+	default:
+		return ErrAccessTypeRequired
+	}
+
+	devicePath, err := i.p.GetDevicePath(
+		ctx, req.VolumeId, req.PublishVolumeInfo, req.UserCredentials)
+	if err != nil {
+		return err
+	}
+	if devicePath == "" {
+		return status.Errorf(
+			codes.FailedPrecondition,
+			"failed to discover device path for volume: %s", req.VolumeId)
+	}
+	if i.validateDevice {
+		validatedDevPath, err := validateDevice(devicePath)
+		if err != nil {
+			return status.Error(codes.FailedPrecondition, err.Error())
+		}
+		devicePath = validatedDevPath
+	}
+	lf["devicePath"] = devicePath
+
+	privMntTgtName, err := i.getPrivateMountTargetName(
+		ctx, req.VolumeId, req.UserCredentials)
+	if err != nil {
+		return err
+	}
+	privMntTgtPath := path.Join(i.privateMountDir, privMntTgtName)
+	lf["privateMountTargetPath"] = privMntTgtPath
+
+	mountTable, err := gofsutil.GetMounts(ctx)
+	if err != nil {
+		return err
+	}
+
+	var (
+		isPub     bool
+		isPrivPub bool
+	)
+
+	for _, m := range mountTable {
+		// Check to see if the device is already mounted to the private
+		// target path.
+		if m.Source == devicePath && m.Path == privMntTgtPath {
+			stat, _ := os.Stat(privMntTgtPath)
+			if stat == nil {
+				continue
+			}
+			if (isBlock && stat.IsDir()) || !stat.IsDir() {
+				return fmt.Errorf(
+					"invalid existing private mount target: %s",
+					privMntTgtPath)
+			}
+			isPrivPub = true
+			lf["deviceToPrivate"] = isPrivPub
+		}
+
+		// Check to see if the private target path is already mounted to the
+		// target path.
+		if m.Source == privMntTgtPath && m.Path == req.TargetPath {
```

Comment:
I know this doesn't work for block.

Reply:
Hmm. Really? I thought since I'm setting Source to the source of the bind mount it would. Isn't the Device field what is set to something like `devtmpfs`?